### PR TITLE
Reconfigure Components

### PR DESCRIPTION
**Pull request by domenicbove**
# Description

This PR enables the running of all.yml for reconfiguration. Previously it is only supported for greenfield installs

- Creates playbooks for each component that front load the host ordering logic as follows:
If component not running, install in parallel
If component running, install in rolling fashion. In the case of zk/kafka, make sure leader/controller goes last

- Removes rolling restart logic from the roles
- Small idempotency fix on connect hub install task
- Changes the logic around creating ca and keystores such that if the files exist, by default they will not get regenerated. If they DNE then they will get created
- Adds tag for package installation, which allows for skipping of those tasks, saving heaps off the total reconfigure time
- Adds configurable wait time after restart to ensure metrics are reporting
- Adds a disabled by default user prompt for rolling restart to pause after each node
- Updated debian and ubuntu dockerfiles to use install latest package

Fixes # (issue)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- Recreating certs without ca, rolling restart functions w zero downtime. Added test `reconfigure-certs.sh` 
- Added `reconfigure-basic.sh` script which tests basic property add
- `all.yml --skip-tags=package` works and saves lots of time

Discovered this can be done: 
```
molecule converge -s rbac-mtls-rhel -- --tags zookeeper --skip-tags package
```

Huge time savings on development!

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] Any variable changes have been validated to be backwards compatible

### CODE REVIEW

File: roles/confluent.variables/defaults/main.yml [hunk 1]
```diff
@@ -1069,3 +1069,21 @@ control_center_health_check_user: "{{control_center_ldap_user}}"
 
 ### Password for authenticated Control Center Health Check. Set if using customized security like Basic Auth.
 control_center_health_check_password: "{{control_center_ldap_password}}"
+
+### Reconfiguration pattern. Set to parallel to reconfigure all hosts at once. May cause downtime
+reconfiguration_pattern: serial
+
+zookeeper_reconfiguration_pattern: "{{reconfiguration_pattern}}"
```
Comment (domenicbove) [inline roles/confluent.variables/defaults/main.yml:1076]: TODO add documentation

File: tasks/create_ordered_kafka_groups.yml [hunk 2]
```diff
@@ -10,10 +10,11 @@
     src: "{{ kafka_broker.config_file }}"
   register: slurped_properties
 
+# TODO think thru the upgrade path, if secrets protection is enabled, but the zookeeper tls client properties file DNE
 - name: Get Controller Broker ID
   shell: |
     {{ binary_base_path }}/bin/zookeeper-shell {{ groups['zookeeper'][0] }}:{{zookeeper_client_port}} \
-      {% if 'zookeeper.ssl.client.enable=true' in slurped_properties.content|b64decode %}--zk-tls-config-file {{kafka_broker.config_file}}{% endif %} \
+      {% if 'zookeeper.ssl.client.enable=true' in slurped_properties.content|b64decode %}-zk-tls-config-file {{kafka_broker.config_file}}{% endif %} \
```
Comment (domenicbove) [inline tasks/create_ordered_kafka_groups.yml:17]: if secrets protection is enabled then kafka_broker.config_file wont work...

File: VARIABLES.md [hunk 3]
```diff
@@ -256,15 +256,15 @@ Default:  "{{ false if ssl_provided_keystore_and_truststore|bool or ssl_custom_c
 
 Boolean to have reruns of all.yml regenerate the certificate authority used for self signed certs
```
Comment (JumaX) [inline VARIABLES.md:257]:
```suggestion
Boolean to have reruns of all.yml regenerate the certificate authority used for self signed certs.
```

File: hosts_example.yml [hunk 4]
```diff
@@ -54,7 +54,7 @@ all:
     # ssl_signed_cert_filepath: "/tmp/certs/{{inventory_hostname}}-signed.crt" # Can be a full chain of certs
     # ssl_key_filepath: "/tmp/certs/{{inventory_hostname}}-key.pem"
     # ssl_key_password: <password for key for each host, will be inputting in the form -passin pass:{{ssl_key_password}} >
-    # regenerate_keystore_and_truststore: true # Set to false once installation is finished to avoid triggering restarts.
+    # regenerate_keystore_and_truststore: true # Set to true to update certs on hosts. If keystores/truststores exist, they won't be updated without this variable
```
Comment (JumaX) [inline hosts_example.yml:57]:
```suggestion
    # regenerate_keystore_and_truststore: true # Set to true to update certs on hosts. If keystores/truststores exist, they won't be updated without this variable.
```

File: kafka_broker.yml [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+  gather_facts: false
+  tags: kafka_broker
+  environment: "{{ proxy_env }}"
+  vars_files:
+    - roles/confluent.kafka_broker/defaults/main.yml
+  tasks:
+    - import_role:
+        name: confluent.variables
+
+    - name: Get Controller Broker ID
```
Comment (JumaX) [inline kafka_broker.yml:44]: Should we look at getting the controller ID via the REST endpoint?  With zookeeper going away sooner then later, would make sense to do this work now. https://confluentinc.atlassian.net/wiki/spaces/KAFKA/pages/929205481/Rest+API+in+the+Broker+to+expose+its+cluster+id

File: roles/confluent.common/tasks/debian.yml [hunk 6]
```diff
@@ -10,6 +10,9 @@
   shell: apt-get install apt-transport-https -y --force-yes
```
Comment (JumaX) [inline roles/confluent.common/tasks/debian.yml:10]: Should we use the APT module here?  It supports forcing installs.

File: tasks/certificate_authority.yml [hunk 7]
```diff
@@ -2,91 +2,144 @@
...
+        mode: 0755
+
+    - name: Create Certificate Authority
+      tags: certificate_authority
+      shell: |
+        openssl req -new -x509 \
+          -keyout /var/ssl/private/generation/{{ssl_self_signed_ca_key_filename}} \
+          -out /var/ssl/private/generation/{{ssl_self_signed_ca_cert_filename}} \
+          -days 365 \
+          -subj '/CN=ca1.test.confluent.io/OU=TEST/O=CONFLUENT/L=PaloAlto/S=Ca/C=US' \
```
Comment (JumaX) [inline tasks/certificate_authority.yml:84]:
```suggestion
          -subj '/CN=ca1.test.confluent.io/OU=TEST/O=CONFLUENT/L=MountainView/S=Ca/C=US' \
```

**Review by JumaX — CHANGES_REQUESTED** [review]
@domenicbove amazing work!  Some really clever changes here. Definitely learned a few things reviewing this.  The biggest thing that jumps out at me, is using Zookeeper to get the controller ID.  Thinking we should be moving away from zookeeper as much as possible in all new features, as it will save us work down the road.

File: kafka_broker.yml (hunk 5 above)
Comment (JumaX) [reply kafka_broker.yml:44]: Understood.  Something else to add to our list in the future. :)

File: roles/confluent.common/tasks/debian.yml (hunk 6 above)
Comment (domenicbove) [reply roles/confluent.common/tasks/debian.yml:10]: will try

File: kafka_broker.yml (hunk 5 above)
Comment (domenicbove) [reply kafka_broker.yml:44]: According to that blog, can also check the `metadata.properties` file, I'll investigate

File: roles/confluent.kafka_broker/tasks/health_check.yml [hunk 8]
```diff
@@ -6,7 +6,7 @@
   register: topics
```
Comment (domenicbove) [inline roles/confluent.kafka_broker/tasks/health_check.yml:6]: TODO add wait task
```
# TODO set default to 60 seconds
- name: Sleep for 300 seconds and continue with play
  wait_for:
    timeout: 60
    # timeout: "{{kafka_broker_startup_wait_time}}"
```

File: kafka_broker.yml (hunk 5 above)
Comment (domenicbove) [reply kafka_broker.yml:44]: lookey lookey:
```
[root@kafka-broker1 data]# cat /var/lib/kafka/data/meta.properties
#
#Tue Jan 05 17:39:43 UTC 2021
cluster.id=qIDfNCw_Sr-fIAT37l9S_A
version=0
broker.id=1
```

Comment (domenicbove) [reply kafka_broker.yml:44]: validated with 2 datadirs there will be a meta.properties in both, containing the same clusterid
```
[root@kafka-broker1 data1]# grep cluster /var/lib/kafka/data1/meta.properties
cluster.id=-5ZcM36rTLWmsy3fHxRxYA
[root@kafka-broker1 data1]# grep cluster /var/lib/kafka/data2/meta.properties
cluster.id=-5ZcM36rTLWmsy3fHxRxYA
```

Comment (domenicbove) [reply kafka_broker.yml:44]: Wow I just realized somehow we got from controller id to cluster id! There is clearly an easy way to get the clusterid.... but thats not what this task is populating! I'm going to leave a comment to the KIP saying how to update when ZK is removed

File: roles/confluent.common/tasks/debian.yml (hunk 6 above)
Comment (domenicbove) [reply roles/confluent.common/tasks/debian.yml:10]: So I just tried and the apt module works without the force flag with our debian image...
I don't want to make a change thats too destructive but I think we are fine to just replace this task with 
```
- name: Install apt-transport-https
  apt:
    name: apt-transport-https
  tags: package
```

The reason I had it as the shell module was a previous debian image (in vagrant) i used did not work with the apt module.

Wondering your thoughts?

File: kafka_broker.yml (hunk 5 above)
Comment (JumaX) [reply kafka_broker.yml:45]: This is really good to know, either way!  Will mark this as resolved, and we can address in the future.  Still a really good tip about the cluster ID though.

File: roles/confluent.common/tasks/debian.yml (hunk 6 above)
Comment (JumaX) [reply roles/confluent.common/tasks/debian.yml:10]: Yeah I think we should change it to the APT module, it's not that big of a change IMO.

Comment (domenicbove) [reply roles/confluent.common/tasks/debian.yml]: ok done

File: roles/confluent.variables/defaults/main.yml [hunk 9]
```diff
@@ -123,11 +123,11 @@ ssl_self_signed_ca_cert_filename: snakeoil-ca-1.crt
...
 ssl_self_signed_ca_password: capassword123
 
-### Boolean to have reruns of all.yml regenerate the certificate authority used for self signed certs
-regenerate_ca: true
+### Boolean to have reruns of all.yml regenerate the certificate authority used for self signed certs.
+regenerate_ca: false
 
-### Boolean to have reruns of all.yml recreate Keystores. Consider disabling this once installation is completed, as this triggers restarts.
-regenerate_keystore_and_truststore: true
+### Boolean to have reruns of all.yml recreate Keystores. On first install, keystores will get created.
```
Comment (JumaX) [inline roles/confluent.variables/defaults/main.yml:129]:
```suggestion
### Boolean to have reruns of all.yml recreate Keystores. On first install, keystores will be created.
```

**Review by JumaX — CHANGES_REQUESTED** [review]
Looks good, just a minor change.

**Review by JumaX — APPROVED** [review]
LGTM.

**Pull request merged**